File: apps/dsa-web/src/api/usage.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export type UsagePeriod = 'today' | 'month' | 'all';

export type UsageCallTypeBreakdown = {
  callType: string;
  calls: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
};

export type UsageModelBreakdown = {
  model: string;
  calls: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  maxTotalTokens: number;
};

export type UsageCallRecord = {
  id: number;
  calledAt: string;
  callType: string;
  model: string;
  stockCode?: string | null;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
};

/** 当前视图作用域：本人用量 vs 跨用户平台聚合。 */
export type UsageScope = 'self' | 'platform';

export type UsageOwnerBreakdown = {
  /** 归属用户 id；平台/后台消耗合并条目为 null。 */
  userId: number | null;
  nickname: string | null;
  ownerScope: 'user' | 'global';
  calls: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  lastCalledAt: string | null;
};

export type UsageByUser = {
  period: UsagePeriod;
  fromDate: string;
  toDate: string;
  scope: UsageScope;
  owners: UsageOwnerBreakdown[];
};

export type UsageDashboard = {
  period: UsagePeriod;
  fromDate: string;
  toDate: string;
  /** 由后端回传，前端据此标注当前是本人视图还是全平台聚合。 */
  scope: UsageScope;
  totalCalls: number;
  totalPromptTokens: number;
  totalCompletionTokens: number;
  totalTokens: number;
  byCallType: UsageCallTypeBreakdown[];
  byModel: UsageModelBreakdown[];
  recentCalls: UsageCallRecord[];
};

export const usageApi = {
  /**
   * 拉取 Token 用量看板。
   *
   * 多用户隔离：`scope: 'platform'` 走 `/api/v1/usage/dashboard`（跨全体用户聚合，
   * 需 `usage.read`，仅运营/管理员）；`scope: 'self'`（默认）走 `/api/v1/usage/me/dashboard`，
   * 只返回当前登录用户自己的用量，普通成员可见。
   */
  getDashboard: async (
    params: { period?: UsagePeriod; limit?: number; scope?: UsageScope } = {},
  ): Promise<UsageDashboard> => {
    const path =
      params.scope === 'platform' ? '/api/v1/usage/dashboard' : '/api/v1/usage/me/dashboard';
    const response = await apiClient.get<Record<string, unknown>>(path, {
      params: {
        period: params.period ?? 'month',
        limit: params.limit ?? 50,
      },
    });

    return toCamelCase<UsageDashboard>(response.data);
  },

  /**
   * 平台级「按用户」用量下钻，需 `usage.read`（运营/管理员）。
   *
   * 定时任务、大盘复盘与后台扇出等不归属任何用户的消耗会合并为一条
   * `ownerScope: 'global'` 条目，不会摊到具体用户头上。
   */
  getByUser: async (
    params: { period?: UsagePeriod; limit?: number } = {},
  ): Promise<UsageByUser> => {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/usage/by-user', {
      params: {
        period: params.period ?? 'month',
        limit: params.limit ?? 100,
      },
    });

    return toCamelCase<UsageByUser>(response.data);
  },
};

File: api/v1/schemas/usage.py
# -*- coding: utf-8 -*-
"""Schemas for LLM usage tracking API."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class CallTypeBreakdown(BaseModel):
    call_type: str = Field(..., description="'analysis' | 'agent' | 'market_review'")
    calls: int
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int


class ModelBreakdown(BaseModel):
    model: str
    calls: int
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int
    max_total_tokens: int = 0


class UsageCallRecord(BaseModel):
    id: int
    called_at: str = Field(..., description="ISO datetime string")
    call_type: str
    model: str
    stock_code: Optional[str] = None
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


class OwnerUsageBreakdown(BaseModel):
    """Per-user token consumption for the platform-level admin drill-down."""

    user_id: Optional[int] = Field(
        None,
        description="Owning user id; null for the aggregated platform bucket.",
    )
    nickname: Optional[str] = Field(None, description="User display name when available")
    owner_scope: str = Field(
        ...,
        description="'user' for a single user, 'global' for platform/background usage",
    )
    calls: int
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int
    last_called_at: Optional[str] = Field(None, description="ISO datetime string")


class UsageSummaryResponse(BaseModel):
    period: str = Field(..., description="'today' | 'month' | 'all'")
    from_date: str = Field(..., description="ISO date string")
    to_date: str = Field(..., description="ISO date string")
    scope: str = Field(
        ...,
        description="'self' for the caller's own usage, 'platform' for the cross-user aggregate",
    )
    total_calls: int
    total_prompt_tokens: int = 0
    total_completion_tokens: int = 0
    total_tokens: int
    by_call_type: List[CallTypeBreakdown]
    by_model: List[ModelBreakdown]


class UsageDashboardResponse(UsageSummaryResponse):
    recent_calls: List[UsageCallRecord]


class UsageByUserResponse(BaseModel):
    """Platform-level per-user usage drill-down (requires ``usage.read``)."""

    period: str = Field(..., description="'today' | 'month' | 'all'")
    from_date: str = Field(..., description="ISO date string")
    to_date: str = Field(..., description="ISO date string")
    scope: str = Field("platform", description="Always 'platform' for this view")
    owners: List[OwnerUsageBreakdown]
